File: 207 (Course Schedule)/java/Solution.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        ArrayList<Integer>[] adj_list = new ArrayList[numCourses];
        for(int i = 0;i<numCourses;i++){
            adj_list[i] = new ArrayList<Integer>();
        }
        int[] degree = new int[numCourses];
        
        for(int[] val: prerequisites){
            adj_list[val[1]].add(val[0]);
            degree[val[0]]++;
        }

        ArrayList<Integer> arr = new ArrayList<>();
        for(int i =0;i<numCourses;i++){
            if (degree[i]==0){
                arr.add(i);
            }   
        }

        for(int i = 0;i<arr.size();i++){
            for(int child: adj_list[arr.get(i)]){
                degree[child]--;
                if(degree[child]==0) arr.add(child);
            }
        }
        return arr.size()==numCourses;
    }
}

File: 207 (Course Schedule)/javascript/Solution.js
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
  let adj_list = new Array(numCourses);
  let degree = new Array(numCourses);
  for(let i =0;i<numCourses;i++){
      degree[i] = 0;
      adj_list[i] = [];
  }

  for(const courses of prerequisites){
      adj_list[courses[1]].push(courses[0]);
      degree[courses[0]]++;
  }

  let validCourses=  []
  for(let i =0 ;i<numCourses;i++){
      if(degree[i]==0){
          validCourses.push(i);
      }
  }

  for(const val of validCourses){
      for(const child of adj_list[val]){
          degree[child]--;
          if (degree[child]==0) validCourses.push(child);
      }
  }
  return validCourses.length==numCourses;
};